File: Trees/Easy/Inorder Traversal/inorder.cpp
#include <iostream>
#include <memory>

// Definition for a binary tree node
class Node {
public:
    int data;
    Node* left = nullptr;
    Node* right = nullptr;

    // Node constructor initializes data
    Node(int val) : data(val) {}
};

// Recursive Inorder Traversal
void inorder(Node* root) {
    if (root != nullptr) {
        // Traverse the left subtree
        inorder(root->left);

        // Print the value of the current node
        std::cout << root->data << " ";

        // Traverse the right subtree
        inorder(root->right);
    }
}

// Function to build a binary tree interactively based on user input
Node* buildTree() {
    int rootData;
    std::cout << "Enter the value for the current node: ";
    std::cin >> rootData;

    auto root = new Node(rootData);

    char hasLeftChild, hasRightChild;
    
    // Ask user if the current node has a left child
    std::cout << "Does the current node have a left child? (y/n): ";
    std::cin >> hasLeftChild;

    // If yes, recursively build the left subtree
    if (hasLeftChild == 'y') {
        root->left = buildTree();
    }

    // Ask user if the current node has a right child
    std::cout << "Does the current node have a right child? (y/n): ";
    std::cin >> hasRightChild;

    // If yes, recursively build the right subtree
    if (hasRightChild == 'y') {
        root->right = buildTree();
    }

    return root;
}

int main() {
    try {
        std::cout << "Interactive Binary Tree Construction\n";
        Node* root = buildTree();

        std::cout << "\nInorder Traversal: ";
        inorder(root);
        std::cout << std::endl;
    } catch (const std::exception& e) {
        // Handle exceptions, if any
        std::cerr << "Error: " << e.what() << std::endl;
    }

    return 0;
}

File: Trees/Easy/Inorder Traversal/inorder.py
# Definition for a binary tree node
class Node:
    def __init__(self, data):
        self.data = data
        self.left = self.right = None

# Recursive Inorder Traversal.
def inorder(root):
    # Check if the root exists
    if root:
        # Perform an inorder traversal on the left subtree
        inorder(root.left)
        # Print the value of the current node
        print(root.data, end=" ")
        # Perform an inorder traversal on the right subtree
        inorder(root.right)

if __name__ == "__main__":
    try:
        # Prompt the user to enter values for the binary tree
        root_data = int(input("Enter the value for the root node: "))
        root = Node(root_data)

        left_data = int(input("Enter the value for the left child of the root node: "))
        root.left = Node(left_data)

        right_data = int(input("Enter the value for the right child of the root node: "))
        root.right = Node(right_data)

        left_left_data = int(input("Enter the value for the left child of the left child: "))
        root.left.left = Node(left_left_data)

        left_right_data = int(input("Enter the value for the right child of the left child: "))
        root.left.right = Node(left_right_data)

        right_left_data = int(input("Enter the value for the left child of the right child: "))
        root.right.left = Node(right_left_data)

        right_right_data = int(input("Enter the value for the right child of the right child: "))
        root.right.right = Node(right_right_data)

        # Perform Inorder Traversal
        print("Inorder Traversal:")
        inorder(root)
    
    except ValueError:
        print("Error: Please enter valid integer values for the nodes.")
    except Exception as e:
        print(f"Error: {e}")
